File: Simulation/Core/ContentDb.cs
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace HistoryTeller.Simulation
{
    /// <summary>
    /// Загруженный контент эпохи. Принимает JSON-строки — источник (TextAsset, Addressables, файл)
    /// решается на уровне Unity, движок об этом не знает.
    /// </summary>
    public sealed class ContentDb
    {
        public Dictionary<string, CharacterDef> Characters { get; private set; }
        public Dictionary<string, SceneDef> Scenes { get; private set; }
        public List<RuleDef> RulesByPriorityDesc { get; private set; }

        public static ContentDb FromJson(string charactersJson, string scenesJson, string rulesJson)
        {
            return new ContentDb
            {
                Characters = JsonConvert.DeserializeObject<List<CharacterDef>>(charactersJson)
                    .ToDictionary(c => c.Id),
                Scenes = JsonConvert.DeserializeObject<List<SceneDef>>(scenesJson)
                    .ToDictionary(s => s.Id),
                RulesByPriorityDesc = JsonConvert.DeserializeObject<List<RuleDef>>(rulesJson)
                    .OrderByDescending(r => r.Priority)
                    .ToList()
            };
        }

        public static LevelDef LoadLevel(string levelJson)
        {
            var obj = JObject.Parse(levelJson);
            var level = new LevelDef
            {
                Id = (string)obj["id"],
                Order = (int?)obj["order"] ?? 0,
                Title = (string)obj["title"],
                Epoch = (string)obj["epoch"],
                Panels = (int)obj["panels"],
                Scenes = obj["scenes"].ToObject<List<string>>(),
                Characters = obj["characters"].ToObject<List<string>>(),
                InitialText = (string)obj["initialText"],
                Goal = GoalNode.Parse((JObject)obj["goal"])
            };
            if (obj["initialState"] != null)
                level.InitialState = obj["initialState"].ToObject<InitialStateDef>();
            if (obj["factCard"] != null)
                level.FactCard = obj["factCard"].ToObject<FactCard>();
            if (obj["solution"] != null)
            {
                level.Solution = new List<Panel>();
                foreach (var t in obj["solution"])
                    level.Solution.Add(new Panel(
                        (string)t["scene"],
                        t["characters"].ToObject<List<string>>().ToArray()));
            }
            return level;
        }
    }
}

File: ios/Modules/Simulation/Sources/Simulation/ContentDb.swift
import Foundation

/// Загруженный контент эпохи. Принимает JSON-строки — источник (бандл, файл, CDN)
/// решается на уровне приложения, движок об этом не знает.
public final class ContentDb {
    public let characters: [String: CharacterDef]
    public let scenes: [String: SceneDef]
    public let rulesByPriorityDesc: [RuleDef]

    init(characters: [String: CharacterDef], scenes: [String: SceneDef], rules: [RuleDef]) {
        self.characters = characters
        self.scenes = scenes
        self.rulesByPriorityDesc = rules
    }

    public static func fromJson(charactersJson: String, scenesJson: String, rulesJson: String) throws -> ContentDb {
        let decoder = JSONDecoder()
        let chars = try decoder.decode([CharacterDef].self, from: Data(charactersJson.utf8))
        let scenes = try decoder.decode([SceneDef].self, from: Data(scenesJson.utf8))
        let rules = try decoder.decode([RuleDef].self, from: Data(rulesJson.utf8))
        return ContentDb(
            characters: Dictionary(chars.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a }),
            scenes: Dictionary(scenes.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a }),
            // sorted(by:) в Swift стабилен → порядок равных приоритетов сохраняется, как в LINQ OrderByDescending.
            rules: rules.sorted { $0.priority > $1.priority }
        )
    }

    public static func loadLevel(_ levelJson: String) throws -> LevelDef {
        try JSONDecoder().decode(LevelDef.self, from: Data(levelJson.utf8))
    }
}
